ergm.py

```diff
@@ -136,72 +142,166 @@ def approximate_auto_correlation_function(self, networks_sample: np.ndarray) ->
+                    is_positive = True
+                else:
+                    first_pos_def_idx += 1
+
+            # Now we find the farthest idx after first_pos_def_idx fir which the sequence of determinants is strictly
+            # monotonically increasing.
+            # TODO: this try-catch intends to catch a bug that hasn't reproduced, consider removing
+            try:
+                # TODO: it might be faster to run in a loop and each time calculate a single determinant, and stop when it
+                #  is smaller than the previous one, rather than computing all of them upfront.
+                determinants = np.linalg.det(possible_cov_mat_ests[first_pos_def_idx:])
+            except IndexError:
+                print(f"shape of features_of_net_samples: {features_of_net_samples.shape}\n"
+                      f"shape of auto_corr_funcs: {auto_corr_funcs.shape}\n"
+                      f"shape of capital_gammas: {capital_gammas.shape}\n"
+                      f"shape of possible_cov_mat_ests: {possible_cov_mat_ests.shape}\n"
+                      f"first_post_def_idx: {first_pos_def_idx}")
+                raise
+            tail_cut_idx = first_pos_def_idx + np.where(np.diff(determinants) < 0)[0][0]
+
+            return possible_cov_mat_ests[tail_cut_idx]
+
         else:
             raise ValueError(f"{method} is an unsupported method for covariance matrix estimation")
 
-    def _calculate_optimization_step(self, observed_features, features_of_net_samples, optimization_method):
-        num_of_features = self._network_statistics.num_of_features
-
+    @staticmethod
+    # @njit
+    def _calculate_optimization_step(observed_features, features_of_net_samples, optimization_method):
         mean_features = np.mean(features_of_net_samples, axis=1)
 
         nll_grad = mean_features - observed_features
 
         if optimization_method == "gradient_descent":
             nll_hessian = None
-            
-        elif optimization_method == "newton_raphson":
-            # # An outer product of the means (E[gi]E[gj])
-            cross_prod_mean_features = (mean_features.reshape(num_of_features, 1) @
-                                        mean_features.T.reshape(1, num_of_features))
-            # A mean of the outer products of the sample (E[gi*gj])
-            mean_features_cross_prod = np.mean(
-                features_of_net_samples.T.reshape(self.sample_size, num_of_features, 1) @
-                features_of_net_samples.T.reshape(self.sample_size, 1, num_of_features), axis=0)
 
-            nll_hessian = mean_features_cross_prod - cross_prod_mean_features
-        else: 
-            raise ValueError(f"Optimization method {optimization_method} not defined") # TODO - throw this error in fit()
+        elif optimization_method == "newton_raphson":
+            # TODO: mean_features is calculated again inside this method, so it can be spared, but it makes more sense
```

Comment:
Only `naive` uses `mean_features` , right?

we could pass both `mean_features` & `features_of_net` to the covariance_matrix_estimation